### PR TITLE
Add time based 2.5D WFS plane wave driving function.

### PR DESCRIPTION
**Pull request by chris-hld**
Major improvements to time domain implementation.

### CODE REVIEW

File: sfs/time/drivingfunction.py [hunk 1]
```diff
@@ -1,64 +1,197 @@
...
+        Amplitude weighting factor for each channel.
+    signal : (N,) array_like
+        Excitation signal (mono) which gets weighted and delayed.
+    fs: int, optional
+        Sampling frequency in hertz.
+
+    Returns
+    -------
+    (N, C) numpy.ndarray
+        N is signal samples(t) per channel C, delayed and weighted.
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: I don't understand this sentence. What does "t" mean?

I think you should move the explanation of C and N up, because those are used before.

File: sfs/time/drivingfunction.py [hunk 2]
```diff
@@ -1,64 +1,197 @@
...
-def delayline(sig, delay, weight):
-    """Delayline with weights, no fractional delay yet.
+def apply_delays(signal, delays, fs=None):
+    """Apply Delays for every channel.
 
-    Delay in samples.
-    sig (m x n) where columns n are channels and rows m the corresponding
-    signals.
-    Weight is array with weighting for each channel.
+    Delay in seconds, no fractional delay. An absolute time offset is applied.
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: You mean "... is returned."?

File: sfs/time/drivingfunction.py [hunk 3]
```diff
@@ -1,64 +1,197 @@
...
-    delay = np.rint(delay).astype(int)  # no fractional delay, samples
-    extention = np.max(delay)  # avoid cutting signal
-    channels = len(weight)  # extract no. of channels
-    sig_length = sig.ndim
-    out = np.zeros([sig_length+extention, channels])  # create shape
+    Parameters
+    ----------
+    signal : (N,) array_like
+        Excitation signal (mono) which gets weighted and delayed.
+    delay : (C,) array_like, in seconds
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: `delays`

File: sfs/time/drivingfunction.py [hunk 4]
```diff
@@ -1,64 +1,197 @@
...
-    """
-    delay = np.rint(delay).astype(int)  # no fractional delay, samples
-    extention = np.max(delay)  # avoid cutting signal
-    channels = len(weight)  # extract no. of channels
-    sig_length = sig.ndim
-    out = np.zeros([sig_length+extention, channels])  # create shape
+    Parameters
+    ----------
+    signal : (N,) array_like
+        Excitation signal (mono) which gets weighted and delayed.
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: There doesn't seem to be any weighting involved, right?

File: sfs/time/drivingfunction.py [hunk 5]
```diff
@@ -1,64 +1,197 @@
...
+    Parameters
+    ----------
+    delays : (C,) array_like
+        Delay in seconds for each channel, negative values allowed.
+    weights : (C,) array_like
+        Amplitude weighting factor for each channel.
+    signal : (N,) array_like
+        Excitation signal (mono) which gets weighted and delayed.
+    fs: int, optional
+        Sampling frequency in hertz.
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: Hertz

File: sfs/time/drivingfunction.py [hunk 6]
```diff
@@ -1,64 +1,197 @@
...
-    sig_length = sig.ndim
-    out = np.zeros([sig_length+extention, channels])  # create shape
+    Parameters
+    ----------
+    signal : (N,) array_like
+        Excitation signal (mono) which gets weighted and delayed.
+    delay : (C,) array_like, in seconds
+        delay in seconds for each channel, negative values allowed.
+    fs: int, optional
+        Sampling frequency in hertz.
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: Hertz

File: sfs/time/soundfield.py [hunk 7]
```diff
@@ -1,48 +1,57 @@
...
+    t : float
+        Simulation point in time (seconds).
+    grid : triple of array_like
+        The grid that is used for the sound field calculations.
+        See `sfs.util.xyz_grid()`.
+    source: function, optional
+        Source type is a function, returning scalar field.
+        For default, see `sfs.time.source.point()`
+    fs: int, optional
+        Sampling frequency in hertz.
```
Comment (mgeier) [inline sfs/time/soundfield.py]: Hertz

File: sfs/time/drivingfunction.py (hunk 3 above)
Comment (mgeier) [inline sfs/time/drivingfunction.py]: superfluous "in seconds"

File: sfs/time/drivingfunction.py [hunk 8]
```diff
@@ -1,64 +1,197 @@
...
         c = defs.c
-    if fs is None:
-        fs = defs.fs
     x0 = util.asarray_of_rows(x0)
-    nx0 = util.asarray_of_rows(nx0)
+    n0 = util.asarray_of_rows(n0)
+    n = util.asarray_1d(n)
+    xref = util.asarray_1d(xref)
+    g0 = np.sqrt(2 * np.pi * np.linalg.norm(xref - x0, axis=1))
+    delays = inner1d(n, x0)/c
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: This would be nicer with spaces around the `/`

File: sfs/time/drivingfunction.py [hunk 9]
```diff
@@ -1,64 +1,197 @@
...
     ds = x0 - xs
-    r = np.linalg.norm(x0-xs, axis=1)
-    delay = 1/c * r
-    weight = g0/(2*np.pi) * inner1d(ds, nx0) / r**(3/2)
-    sig = np.ones(len(x0))  # dirac
-    line = delayline(sig, delay*fs, weight)
-    return delay, weight, line
+    r = np.linalg.norm(x0 - xs, axis=1)
+    delays = r/c
+    weights = g0/(2 * np.pi) * inner1d(ds, n0) / r**(3/2)
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: You could use the same numerator/denominator as in the docstring, the two fractions here look a bit confusing, don't they?

Or is there a reason for writing it like that?

File: sfs/time/drivingfunction.py [hunk 10]
```diff
@@ -1,64 +1,197 @@
...
+    signal = util.asarray_1d(signal)
+    delays = util.asarray_1d(delays)
+
+    # convert to samples
+    delays_samples = np.rint(fs * delays).astype(int)
+    offset_samples = delays_samples.min()
+    delays_samples -= offset_samples
+    out = np.zeros((delays_samples.max() + len(signal), len(delays_samples)))
+    for channel, cdelay in enumerate(delays_samples):
+        out[cdelay: cdelay + len(signal), channel] = signal
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: superfluous space after colon

File: sfs/time/drivingfunction.py [hunk 11]
```diff
@@ -1,64 +1,197 @@
...
+        Simulation point in time offset (seconds).
 
-    return out
+    """
+    if fs is None:
+        fs = defs.fs
+    signal = util.asarray_1d(signal)
+    delays = util.asarray_1d(delays)
+
+    # convert to samples
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: Since the word `samples` is very frequently used in the following few lines, is this comment really necessary?

File: sfs/time/soundfield.py [hunk 12]
```diff
@@ -1,48 +1,57 @@
...
+
+
+def p_array(x0, d, channel_weights, t, grid, source=point, fs=None, c=None):
+    """Compute sound field for an array of secondary sources for a given time.
+
+    Parameters
+    ----------
+    x0 : (N, 3) array_like
+        Sequence of secondary source positions.
+    d : (N, C) numpy.ndarray
```
Comment (mgeier) [inline sfs/time/soundfield.py]: `array_like`

File: sfs/time/soundfield.py [hunk 13]
```diff
@@ -1,48 +1,57 @@
...
+        For default, see `sfs.time.source.point()`
+    fs: int, optional
+        Sampling frequency in hertz.
+    c : float, optional
+        Speed of sound.
+
+    Returns
+    -------
+    numpy.ndarray
+        Pressure at grid positions.
```
Comment (mgeier) [inline sfs/time/soundfield.py]: Sound pressure

File: sfs/time/soundfield.py [hunk 14]
```diff
@@ -1,48 +1,57 @@
...
+    channel_weights : (C,) array_like
+        Additional weights applied during integration, e.g. source tapering.
+    t : float
+        Simulation point in time (seconds).
+    grid : triple of array_like
+        The grid that is used for the sound field calculations.
+        See `sfs.util.xyz_grid()`.
+    source: function, optional
+        Source type is a function, returning scalar field.
+        For default, see `sfs.time.source.point()`
```
Comment (mgeier) [inline sfs/time/soundfield.py]: .

File: sfs/time/soundfield.py [hunk 15]
```diff
@@ -1,48 +1,57 @@
...
     """
     if fs is None:
         fs = defs.fs
-    channel_weight = util.asarray_1d(channel_weight)
-    if len(channel_weight) != len(x0):
+    if c is None:
+        c = defs.c
+    x0 = util.asarray_of_rows(x0)
+    channel_weights = util.asarray_1d(channel_weights)
+    if np.size(channel_weights, 0) != np.size(x0, 0):
```
Comment (mgeier) [inline sfs/time/soundfield.py]: Every time someone uses `np.size()`, a fairy dies. To avoid this, you should use `len(channel_weights)` and `len(x0)`.

Comment (mgeier) [inline sfs/time/soundfield.py]: The number of columns of `d` must also be the same.

And before killing another fairy, please use `d.shape[1]` or `len(d.T)`, I'm not sure which one is better.

File: sfs/time/soundfield.py [hunk 16]
```diff
@@ -1,48 +1,57 @@
...
+        Pressure at grid positions.
 
     """
     if fs is None:
         fs = defs.fs
-    channel_weight = util.asarray_1d(channel_weight)
-    if len(channel_weight) != len(x0):
+    if c is None:
+        c = defs.c
+    x0 = util.asarray_of_rows(x0)
```
Comment (mgeier) [inline sfs/time/soundfield.py]: You should make sure that `d` is an array:

```python
d = np.asarray(d)
```

Probably now is the time to create a function `asarray_of_columns()`?

File: sfs/time/source.py [hunk 17]
```diff
@@ -2,30 +2,58 @@
...
 from .. import util
 from .. import defs
 
 
-def greens_function_ps(xs, grid, t, fs=None, c=None):
-    """Source model for a point source: 3D Green's function.
+def point(xs, signal, t, grid, fs=None, c=None):
+    r"""Source model for a point source: 3D Green's function.
+
+    Calculates the scalar pressure field for a given point in time, evoked by
```
Comment (mgeier) [inline sfs/time/source.py]: sound pressure

File: sfs/time/source.py [hunk 18]
```diff
@@ -2,30 +2,58 @@
...
+        See `sfs.util.xyz_grid()`.
+    fs: int, optional
+        Sampling frequency in hertz.
+    c : float, optional
+        Speed of sound.
+
+    Returns
+    -------
+    numpy.ndarray
+        Scalar pressure field, evaluated at positions given by *grid*.
```
Comment (mgeier) [inline sfs/time/source.py]: sound pressure

File: sfs/time/source.py [hunk 19]
```diff
@@ -2,30 +2,58 @@
...
-                     4pi |x-xs|
+        g(x-x_s,t) = \frac{1}{4 \pi |x - x_s|} \dirac{t - \frac{|x - x_s|}{c}}
 
     """
     xs = util.asarray_1d(xs)
     grid = util.as_xyz_components(grid)
     if c is None:
         c = defs.c
-    if fs is None:
-        fs = defs.fs
```
Comment (mgeier) [inline sfs/time/source.py]: Why did you remove that?

File: sfs/time/source.py [hunk 20]
```diff
@@ -2,30 +2,58 @@
...
+    -----
+    .. math::
 
-                         1
-         g(x-xs,t) = ---------- delta(t - |x-xs|/c)
-                     4pi |x-xs|
+        g(x-x_s,t) = \frac{1}{4 \pi |x - x_s|} \dirac{t - \frac{|x - x_s|}{c}}
 
     """
     xs = util.asarray_1d(xs)
```
Comment (mgeier) [inline sfs/time/source.py]: You should ensure that `signal` is mono.

File: sfs/time/source.py (hunk 19 above)
Comment (mgeier) [inline sfs/time/source.py]: You can of course remove it, but then you should also remove the default value `fs=None` above.

File: sfs/time/drivingfunction.py [hunk 21]
```diff
@@ -1,64 +1,197 @@
...
+    weights : (C,) array_like
+        Amplitude weighting factor for each channel.
+    signal : (N,) array_like
+        Excitation signal (mono) which gets weighted and delayed.
+    fs: int, optional
+        Sampling frequency in hertz.
+
+    Returns
+    -------
+    (N, C) numpy.ndarray
```
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/drivingfunction.py (hunk 2 above)
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: Also applied, isn't it?

File: sfs/time/drivingfunction.py (hunk 3 above)
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/drivingfunction.py (hunk 4 above)
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: True. In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/drivingfunction.py (hunk 5 above)
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/drivingfunction.py (hunk 6 above)
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/soundfield.py (hunk 7 above)
Comment (chris-hld) [inline sfs/time/soundfield.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/source.py [hunk 22]
```diff
@@ -2,30 +2,58 @@
...
+        Position of source in cartesian coordinates.
+    signal : (N,) array_like
+        Excitation signal.
+    t : float
+        Observed point in time.
+    grid : triple of array_like
+        The grid that is used for the sound field calculations.
+        See `sfs.util.xyz_grid()`.
+    fs: int, optional
+        Sampling frequency in hertz.
```
Comment (chris-hld) [inline sfs/time/source.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/drivingfunction.py (hunk 8 above)
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/drivingfunction.py [hunk 23]
```diff
@@ -1,64 +1,197 @@
...
+    xref = util.asarray_1d(xref)
+    g0 = np.sqrt(2 * np.pi * np.linalg.norm(xref - x0, axis=1))
     ds = x0 - xs
-    r = np.linalg.norm(x0-xs, axis=1)
-    delay = 1/c * r
-    weight = g0/(2*np.pi) * inner1d(ds, nx0) / r**(3/2)
-    sig = np.ones(len(x0))  # dirac
-    line = delayline(sig, delay*fs, weight)
-    return delay, weight, line
+    r = np.linalg.norm(x0 - xs, axis=1)
```
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: Absolutely. In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/drivingfunction.py (hunk 9 above)
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: I thought it's way nicer to read it like that, because of the bit bulky linalg.norm() etc..

File: sfs/time/drivingfunction.py (hunk 10 above)
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/drivingfunction.py (hunk 11 above)
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: No. commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/soundfield.py (hunk 13 above)
Comment (chris-hld) [inline sfs/time/soundfield.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/soundfield.py (hunk 14 above)
Comment (chris-hld) [inline sfs/time/soundfield.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/source.py (hunk 17 above)
Comment (chris-hld) [inline sfs/time/source.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

File: sfs/time/source.py (hunk 18 above)
Comment (chris-hld) [inline sfs/time/source.py]: In commit dc23fc443a07d4103b36d84525725a9a9d0be6e6

**Comment by chris-hld** [comment]
Thank you for that very detailed review!! I'm about to finish the remaining updates.

File: sfs/time/soundfield.py (hunk 15 above)
Comment (chris-hld) [inline sfs/time/soundfield.py]: This causes problems for only one channel, doesn't it?

File: sfs/time/source.py (hunk 20 above)
Comment (chris-hld) [inline sfs/time/source.py]: In commit 8a77179efe033fc4a25c171dd675fc2be2a7b334

File: sfs/time/source.py (hunk 19 above)
Comment (chris-hld) [inline sfs/time/source.py]: in commit 8a77179efe033fc4a25c171dd675fc2be2a7b334

File: sfs/time/soundfield.py (hunk 16 above)
Comment (chris-hld) [inline sfs/time/soundfield.py]: Yes, we could add that!

File: sfs/time/soundfield.py (hunk 14 above)
Comment (chris-hld) [inline sfs/time/soundfield.py]: in commit 8a77179efe033fc4a25c171dd675fc2be2a7b334

File: sfs/time/soundfield.py (hunk 13 above)
Comment (chris-hld) [inline sfs/time/soundfield.py]: in commit 8a77179efe033fc4a25c171dd675fc2be2a7b334

File: sfs/time/drivingfunction.py [hunk 24]
```diff
@@ -1,64 +1,198 @@
...
+        Amplitude weighting factor for each channel.
+    signal : (N,) array_like
+        Excitation signal (mono) which gets weighted and delayed.
+    fs: int, optional
+        Sampling frequency in Hertz.
+
+    Returns
+    -------
+    driving_signals : (N, C) numpy.ndarray
+        Driving signals (with N samples) per channel C.
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: I think you should extend the parentheses until the end of the line, or remove them completely (which might be even better).

File: sfs/time/drivingfunction.py (hunk 2 above)
Comment (mgeier) [inline sfs/time/drivingfunction.py]: I'm not sure. Aren't you talking about the return parameter `delay`?
As far as I understand it, this is not *applied*. If anything, it is *removed* (or *subtracted*) from all the delays in `delays`.
But I'm not a native speaker, so I might be completely wrong.

File: sfs/time/drivingfunction.py (hunk 24 above)
Comment (mgeier) [inline sfs/time/drivingfunction.py]: Looking at it again, I'm not so sure anymore ... but something seems strange here.

Probably it's not necessary to mention `N`, it isn't mentioned above, either.

Also, the plural seems to be inappropriate. I think it's either one signal per channel or one multichannel signal with `C` channels or `C` signals, but not multiple signals per channel.

Same for `apply_delays()` below.

File: sfs/time/drivingfunction.py [hunk 25]
```diff
@@ -1,64 +1,198 @@
...
+        Mono excitation signal (with N samples) which gets delayed.
+    delays : (C,) array_like
+        Delay in seconds for each channel (C), negative values allowed.
+    fs: int, optional
+        Sampling frequency in Hertz.
+
+    Returns
+    -------
+    out : (N, C) numpy.ndarray
+        Output signals (with N samples) per channel C.
```
Comment (mgeier) [inline sfs/time/drivingfunction.py]: See above.

File: sfs/time/soundfield.py (hunk 12 above)
Comment (mgeier) [inline sfs/time/soundfield.py]: Yes, it should indeed be two-dimensional.

File: sfs/time/soundfield.py (hunk 16 above)
Comment (mgeier) [inline sfs/time/soundfield.py]: Do you want to add this to this PR our shall we do it afterwards?

File: sfs/time/soundfield.py (hunk 15 above)
Comment (mgeier) [inline sfs/time/soundfield.py]: No need to use indexing here. `len(x0)` is the length of x0, regardless if its one- or two-dimensional.

But if we had a `asarray_of_columns()` function, this would guarantee that the array is always two-dimensional, even for mono input.

File: sfs/time/soundfield.py [hunk 26]
```diff
@@ -1,48 +1,57 @@
...
-    # zero padding and reverse time signals
-    d_line_inverted = np.flipud(np.row_stack((zero_fill, d_line, zero_fill)))
-
+    if c is None:
+        c = defs.c
+    x0 = util.asarray_of_rows(x0)
+    channel_weights = util.asarray_1d(channel_weights)
+    d = np.asarray(d)
+    if (len(channel_weights) != x0.shape[0]) or (
+            len(channel_weights) != d.shape[1]):
```
Comment (mgeier) [inline sfs/time/soundfield.py]: No need for `or` here, you can just chain the comparisons. Yay Python!

And `x0.shape[0]` is equivalent to `len(x0)`. I would prefer latter.

**Comment by mgeier** [comment]
Thanks for the changes! I hope I could clear up the few remaining points.

I have one more thing: PEP 8 tells us to limit docstring lines to 72 characters:
https://www.python.org/dev/peps/pep-0008/#maximum-line-length

Would be nice if you could also do this, right now some lines are longer.
Of course there may be some exceptions, but in most cases I think it would be good to limit the docstring line length to 72.

File: sfs/time/soundfield.py (hunk 16 above)
Comment (chris-hld) [inline sfs/time/soundfield.py]: I'd like to do this afterwards, because I think I'll need help with that.

Comment (mgeier) [inline sfs/time/soundfield.py]: Sure, no problem.

Just for the record, it could be very similar to `asarray_of_rows()`, there's really only one character moved to a different place:

```python
def asarray_of_columns(a, **kwargs):
    result = np.squeeze(np.asarray(a, **kwargs))
    if result.ndim == 1:
        result = result.reshape(-1, 1)
    return result
```

File: sfs/time/soundfield.py (hunk 26 above)
Comment (chris-hld) [inline sfs/time/soundfield.py]: Yay Python! In commit 6765559c5cc2065e08a2254e0e3d5413fb3edde7

File: sfs/time/drivingfunction.py (hunk 24 above)
Comment (chris-hld) [inline sfs/time/drivingfunction.py]: How do you like this: In 6765559c5cc2065e08a2254e0e3d5413fb3edde7

**Comment by chris-hld** [comment]
I updated the remaining points, however, simple and short docstring explanations become more difficult to me... Feel free to comment further!

**Comment by mgeier** [comment]
Writing good docstrings is indeed hard. But I think yours are already better than those in other submodules. We should continually try to improve all docstrings in the future.

But first, LGTM!

Thanks for the great work!
